src/Carcassonne/Game/Game.cpp

```diff
@@ -518,4 +523,249 @@ std::pair<Direction, int> Game::move_score(Player player, TileType tile_type, Ti
    return std::make_pair(best_dir, total_score);
 }
 
+template <typename E>
+constexpr inline typename std::underlying_type<E>::type to_underlying(E e) noexcept {
+    return static_cast<typename std::underlying_type<E>::type>(e);
+}
+
+template <typename T, typename U>
+inline mb::result<mb::empty> flags_to_states(const T flags, const int neurons_to_set, U output_it) {
+   auto num = to_underlying(flags);
+   // check commented out for speed
+   // if (neurons_to_set != log2(num) + 1)
+   //    return mb::error("invalid neurons_to_set");
+   std::vector<bool> states(neurons_to_set, false);
+   for (int i = 0; num; ++i) {
+       states[i] = num % 2;
+       num /= 2;
+   }
+   reverse_copy(states.begin(), states.end(), output_it);
+   std::advance(output_it, neurons_to_set);
+   return mb::ok;
+}
+
+template <typename T, typename U>
+inline mb::result<mb::empty> vector_to_states(const T vector, const size_t neurons_to_set, U& output_it) {
+   // check commented out for speed
+   if (neurons_to_set % vector.size())
+      return mb::error("invalid neurons_to_set");
+   std::vector<bool> states(neurons_to_set, false);
+   for (int i = 0; i < vector.size(); ++i) {
+      auto val = to_underlying(vector.at(i));
+       states.at(val + neurons_to_set / vector.size() * i) = true;
+   }
+   copy(states.begin(), states.end(), output_it);
+   std::advance(output_it, neurons_to_set);
+   return mb::ok;
+}
+
+template <typename U>
+inline mb::result<mb::empty> tile_to_caffe(const Tile& tile, U& output_it) {
```

Comment:
W sumie to te funkcje nie zwracają błędów

Reply:
Chciałem, żeby zwracały błędy jak potem będziemy zmieniać ilości neuronów na cechę i źle się dobierze pary "neurons_to_set, vector.size()" i "neurons_to_set i ilość bajtów we fladze" ale masz racje, jest to nie potrzebne, checki tylko spowalniają program